src/relations/database_provides.py

```diff
@@ -1,142 +1,172 @@
+        """Whether database user has been shared with application charm"""
+        for key in ["database", "username", "password", "endpoints"]:
+            if key not in self._local_databag:
+                return False
+        return True
+
+    @property
+    def _database(self) -> str:
+        """Requested database name"""
+        return self._remote_databag["database"]
+
+    def _get_username(self, database_requires_username: str) -> str:
+        """Database username"""
+        # Prefix username with username from database requires relation.
+        # This ensures a unique username if MySQL Router is deployed in a different Juju model
+        # from MySQL.
+        # (Relation IDs are only unique within a Juju model.)
+        return f"{database_requires_username}-{self.id}"
+
+    def _set_databag(self, *, username: str, password: str, router_endpoint: str) -> None:
+        """Share connection information with application charm."""
+        read_write_endpoint = f"{router_endpoint}:6446"
+        read_only_endpoint = f"{router_endpoint}:6447"
+        logger.debug(
+            f"Setting databag {self.id=} {self._database=}, {username=}, {read_write_endpoint=}, {read_only_endpoint=}"
         )
-        self.framework.observe(
-            self.charm.on[PEER].relation_changed, self._on_peer_relation_changed
+        self._interface.set_database(self.id, self._database)
+        self._interface.set_credentials(self.id, username, password)
+        self._interface.set_endpoints(self.id, read_write_endpoint)
+        self._interface.set_read_only_endpoints(self.id, read_only_endpoint)
+        logger.debug(
+            f"Set databag {self.id=} {self._database=}, {username=}, {read_write_endpoint=}, {read_only_endpoint=}"
         )
 
-        self.framework.observe(
-            self.charm.on[DATABASE_PROVIDES_RELATION].relation_broken, self._on_database_broken
-        )
-
-    # =======================
-    #  Handlers
-    # =======================
-
-    def _on_database_requested(self, event: DatabaseRequestedEvent) -> None:
-        """Handle the database requested event."""
-        if not self.charm.unit.is_leader():
-            return
-
-        # Store data in databag to trigger DatabaseRequires initialization in database_requires.py
-        self.charm.app_peer_data[MYSQL_ROUTER_PROVIDES_DATA] = json.dumps(
-            {"database": event.database, "extra_user_roles": event.extra_user_roles}
+    def _delete_databag(self) -> None:
+        """Remove connection information from databag."""
+        logger.debug(f"Deleting databag {self.id=}")
+        self._local_databag.clear()
+        logger.debug(f"Deleted databag {self.id=}")
+
+    def create_database_and_user(self, *, router_endpoint: str, shell: mysql_shell.Shell) -> None:
+        """Create database & user and update databag."""
+        username = self._get_username(shell.username)
+        password = shell.create_application_database_and_user(
```

Comment:
What if `extra_user_roles` is requested?
What if `extra_user_roles=mysqlrouter`?

Reply:
> Does mysqlrouter need to support additional mysqlrouters connected on its DatabaseProvides interface?

We shouldn't support a cascade of routers.
But questions here - what should UX be if a user requests `extra_user_roles`, which are not supported by this code?
Another Q - how is the code around `extra_user_roles` shared between MySQL and mysql-router charms?